walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,124 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.List;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+
+	private static final int END_OF_DAY = 1440;
+
+	public Collection<TimeRange> query(Collection<Event> events,
+		MeetingRequest request) {
+		int duration = (int) request.getDuration();
+		Collection<String> attendees = request.getAttendees();
+		Collection<String> optionalAttendees = request.getOptionalAttendees();
+
+		Collection<TimeRange> availableTimes = new ArrayList<>();
+		Collection<TimeRange> availableTimesWithOptionals = new ArrayList<>();
+
+		availableTimes = getAvaliableTimes(attendees, events, duration);
+
+		if (!optionalAttendees.isEmpty()) {
+			availableTimesWithOptionals = getAvaliableTimes(optionalAttendees,
+				events, duration);
+		}
+
+		if (optionalAttendees.isEmpty()) {
+			return availableTimes;
+		}
+
+		if (attendees.isEmpty()) {
+			return availableTimesWithOptionals;
+		}
+
+		availableTimesWithOptionals = intersection(availableTimesWithOptionals,
+			availableTimes);
+
+		if (!availableTimesWithOptionals.isEmpty()) {
+			return availableTimesWithOptionals;
+		}
+
+		return availableTimes;
+	}
+
+	/**
+	 * Finds avaliable times for the given atteendee list, input 
+	 * collection and event duration
+	 */
+	private Collection<TimeRange> getAvaliableTimes(Collection<String> attendees,
+		Collection<Event> events, int duration) {
+		Collection<TimeRange> availableTimes = new ArrayList<>();
+
+		int start = 0;
+		int end = start + duration;
+		int availablePeriodStart = -1, availablePeriodEnd = -1;
+
+		while (end<= END_OF_DAY) {
+			TimeRange currentTimeRange = TimeRange.fromStartDuration(start, duration);
+			boolean works = true;
+
+			for (Event event: events) {
+				if (!works) break;
+
+				TimeRange eventTimeRange = event.getWhen();
+				if (currentTimeRange.overlaps(eventTimeRange)) {
+					Collection<String> eventAttendees = event.getAttendees();
+					if (!Collections.disjoint(attendees, eventAttendees)) {
+						works = false;
+						continue;
+					}
+				}
+			}
+
+			if (works) {
+				if (availablePeriodEnd == -1) {
+					availablePeriodStart = start;
+				}
+
+				availablePeriodEnd = end;
+
+				if (end == END_OF_DAY) {
+					TimeRange availableTime = TimeRange
+						.fromStartEnd(availablePeriodStart, availablePeriodEnd, false);
+					availableTimes.add(availableTime);
+				}
+			} else {
+				if (availablePeriodEnd != -1) {
+					TimeRange availableTime = TimeRange
+						.fromStartEnd(availablePeriodStart, availablePeriodEnd, false);
+					availableTimes.add(availableTime);
+				}
+
+				availablePeriodStart = -1;
+				availablePeriodEnd = -1;
+			}
+			start++;
+			end++;
+		}
```

Comment:
Some comments about how the algorithm works and the meaning of key variables (like `start`, `end`, `availablePeriodStart`, and `availablePeriodEnd`) would be helpful. I usually put a comment above each loop that describes what we are looping through and why.